lib/fb/insight_request.rb

```diff
@@ -0,0 +1,41 @@
+# Ruby client to authenticate a Facebook user.
+# @see http://www.rubydoc.info/gems/Fb/
+module Fb
+  # @api private
+  class InsightRequest
+    # @param [Hash] options to initialize an insight request.
+    # @option [<String>] :ids the ids of the facebook resources.
+    #   InsightRequest accepts a singular id or an array of ids.
+    # @option [String] :access_token an access token associated with given ids.
+    def initialize(options = {})
+      @params = options.merge metric: options.fetch(:metric).join(',')
+    end
+
+    # Builds all required requests and spawns threads to make calls
+    # asynchronously. Each response is appended synchronously to an array of
+    # responses to maintain thread safety.
+    # @return [<Net::HTTPResponse>] an array of responses.
+    # @raise [Fb::HTTPError] if there is a failed request when joining threads.
+    def run
+      threads, mutex, responses = [], Mutex.new, []
+      requests.each do |request|
+        threads << Thread.new(request, responses) do |request, responses|
+          mutex.synchronize { responses << request.run }
```

Comment:
I can merge `request.run.body` here instead of just the response code, but I think this is easier to test or find failed requests.